Libraries/Microsoft.Teams.Plugins/Microsoft.Teams.Plugins.External/Microsoft.Teams.Plugins.External.McpClient/McpClientPlugin.cs

```diff
@@ -0,0 +1,246 @@
+                    else
+                    {
+                        throw new Exception($"Failed to fetch tools from MCP server at {url}", fetchTask.Exception);
+                    }
+                }
+
+                var tools = fetchTask.Result;
+                if (!Cache.ContainsKey(url))
+                {
+                    Cache[url] = new McpCachedValue();
+                }
+
+                Cache[url].AvailableTools = tools;
+                Cache[url].LastFetched = DateTimeOffset.UtcNow;
+                Cache[url].Transport = pluginParams.Transport;
+
+                _logger.Debug($"Cached {tools.Count} tools from MCP server at {url}");
+            }
+        }
+    }
+
+    private async Task<List<McpToolDetails>> FetchToolsFromServer(Uri url, McpClientPluginParams pluginParams)
+    {
+        IClientTransport transport = CreateTransport(url, pluginParams.Transport, pluginParams.Headers);
+        var client = await McpClientFactory.CreateAsync(transport);
+        var tools = await client.ListToolsAsync();
+
+        // Convert MCP tools to our format
+        var mappedTools = tools.Select(t => new McpToolDetails()
+        {
+            Name = t.Name,
+            Description = t.Description,
+            InputSchema = t.JsonSchema
+        }).ToList();
+
+        return mappedTools;
+    }
+
+    private IClientTransport CreateTransport(Uri url, McpClientTransport transport, IDictionary<string, string>? headers)
+    {
+        var options = new SseClientTransportOptions() { Endpoint = url };
+        switch (transport)
+        {
+            case McpClientTransport.StreamableHttp:
+                options.TransportMode = HttpTransportMode.StreamableHttp;
+                break;
+            case McpClientTransport.Sse:
+                options.TransportMode = HttpTransportMode.Sse;
+                break;
+            default:
+                options.TransportMode = HttpTransportMode.AutoDetect;
+                break;
+        }
+        options.AdditionalHeaders = headers;
+        return new SseClientTransport(options);
+    }
+
+    private AI.Function CreateFunctionFromTool(Uri url, McpToolDetails tool, McpClientPluginParams pluginParams)
+    {
+        return new AI.Function(
+            tool.Name, 
+            tool.Description, 
+            JsonSchema.FromText(tool.InputSchema?.GetRawText() ?? "{}"), 
+            async (IDictionary<string, object?> args) =>
+            {
+                try
+                {
+                    _logger.Debug($"Making call to {url} for tool {tool.Name}");
+                    string result = await CallMcpTool(url, tool, args.AsReadOnly(), pluginParams);
+                    _logger.Debug($"Received result from {tool.Name}: {result}");
+                    return result;
+                }
+                catch (Exception ex)
+                {
+                    _logger.Error($"Error calling MCP tool {tool.Name} at {url}", ex);
+                    throw;
```

Comment:
throw new InvalidOperationException($"Failed to call MCP tool {tool.Name} at {url}", ex);